src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/UiElementExtensions.cs

```diff
@@ -0,0 +1,46 @@
+// This Source Code Form is subject to the terms of the MIT License.
+// If a copy of the MIT was not distributed with this file, You can obtain one at https://opensource.org/licenses/MIT.
+// Copyright (C) Leszek Pomianowski and WPF UI Contributors.
+// All Rights Reserved.
+
+namespace System.Windows.Controls;
+
+internal static class UiElementExtensions
+{
+    /// <summary>
+    /// Do not call it outside of NCHITTEST, NCLBUTTONUP, NCLBUTTONDOWN messages!
+    /// </summary>
+    /// <returns><see langword="true"/> if mouse is over the element. <see langword="false"/> otherwise.</returns>
+    public static bool IsMouseOverElement(this UIElement element, IntPtr lParam)
+    {
+        // This method will be invoked very often and must be as simple as possible.
+        if (lParam == IntPtr.Zero)
+        {
+            return false;
+        }
+
+        try
+        {
+            var mousePosScreen = new Point(Get_X_LParam(lParam), Get_Y_LParam(lParam));
+            var bounds = new Rect(default(Point), element.RenderSize);
+
+            Point mousePosRelative = element.PointFromScreen(mousePosScreen);
+
+            return bounds.Contains(mousePosRelative);
+        }
+        catch
+        {
+            return false;
+        }
+    }
+
+    private static int Get_X_LParam(IntPtr lParam)
+    {
+        return (short)(lParam.ToInt32() & 0xFFFF);
```

Comment:
Should not use ToInt32, please use ToInt64 replace, see https://github.com/dotnet/wpf/issues/6777